src/ray/common/ray_syncer/ray_syncer.cc

```diff
@@ -0,0 +1,391 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/common/ray_syncer/ray_syncer.h"
+
+#include "ray/common/ray_config.h"
+namespace ray {
+namespace syncer {
+
+NodeState::NodeState() { snapshots_taken_.fill(-1); }
+
+bool NodeState::SetComponents(RayComponentId cid,
+                              const ReporterInterface *reporter,
+                              ReceiverInterface *receiver) {
+  if (cid < static_cast<RayComponentId>(kComponentArraySize) &&
+      reporters_[cid] == nullptr && receivers_[cid] == nullptr) {
+    reporters_[cid] = reporter;
+    receivers_[cid] = receiver;
+    return true;
+  } else {
+    RAY_LOG(ERROR) << "Fail to set components, component_id:" << cid
+                   << ", reporter:" << reporter << ", receiver:" << receiver;
+    return false;
+  }
+}
+
+std::optional<RaySyncMessage> NodeState::GetSnapshot(RayComponentId cid) {
+  if (reporters_[cid] == nullptr) {
+    return std::nullopt;
+  }
+  auto message = reporters_[cid]->Snapshot(snapshots_taken_[cid], cid);
+  if (message != std::nullopt) {
+    snapshots_taken_[cid] = message->version();
+    RAY_LOG(DEBUG) << "Snapshot taken: cid:" << cid << ", version:" << message->version()
+                   << ", node:" << NodeID::FromBinary(message->node_id());
+  }
+  return message;
+}
+
+bool NodeState::ConsumeMessage(std::shared_ptr<const RaySyncMessage> message) {
+  auto &current = cluster_view_[message->node_id()][message->component_id()];
+
+  RAY_LOG(DEBUG) << "ConsumeMessage: " << (current ? current->version() : -1)
+                 << " message_version: " << message->version()
+                 << ", message_from: " << NodeID::FromBinary(message->node_id());
+  // Check whether newer version of this message has been received.
+  if (current && current->version() >= message->version()) {
+    return false;
+  }
+
+  current = message;
+  auto receiver = receivers_[message->component_id()];
+  if (receiver != nullptr) {
+    receiver->Update(message);
+  }
+  return true;
+}
+
+NodeSyncConnection::NodeSyncConnection(RaySyncer &instance,
+                                       instrumented_io_context &io_context,
+                                       std::string node_id)
+    : timer_(io_context),
+      instance_(instance),
+      io_context_(io_context),
+      node_id_(std::move(node_id)) {}
+
+bool NodeSyncConnection::PushToSendingQueue(
+    std::shared_ptr<const RaySyncMessage> message) {
+  if (message->node_id() == GetNodeId()) {
```

Comment:
Can you tell me when  this happen ?

Reply:
Let me put some comments about this function.

Reply:
> Actually, it should be covered in the code after this line, but this tries to make sure it's robust and self-complete byitself.

If we think it cannot happen, can we use CHECK instead so future readers don't need to think about under which case this if statement can be triggered.